### PR TITLE
Tool module for CPA-witness-exec

### PR DESCRIPTION
**Pull request by leostrakosch**
Intended for use with scripts/cpa_witness_exec.py of CPAchecker repo to validate violation witnesses through dynamic execution of a created test suite.

Based on cpachecker tool module.

### CODE REVIEW

File: benchexec/tools/cpa-witness-exec.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
+"""
+BenchExec is a framework for reliable benchmarking.
+This file is part of BenchExec.
+
+Copyright (C) 2007-2015  Dirk Beyer
```
Comment (kfriedberger) [inline benchexec/tools/cpa-witness-exec.py]: 2015?

File: benchexec/tools/cpa-witness-exec.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+import sys
+import os
+
+import benchexec.result as result
+import benchexec.util as util
+import benchexec.tools.template
+from benchexec.model import SOFTTIMELIMIT
+
+
+class Tool(benchexec.tools.template.BaseTool):
```
Comment (kfriedberger) [inline benchexec/tools/cpa-witness-exec.py]: There is much in common with the CPAchecker-tool-wrapper.
Would it be possible to merge those two?
Or inherit this from the other tool?

File: benchexec/tools/cpa-witness-exec.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    """
+    Tool info for CPA-witness-exec.
+    It has additional features such as building CPAchecker before running it
+    if executed within a source checkout.
+    It also supports extracting data from the statistics output of CPAchecker
+    for adding it to the result tables.
+    """
+
+    REQUIRED_PATHS = [
```
Comment (kfriedberger) [inline benchexec/tools/cpa-witness-exec.py]: same as CPAchecker-tool-wrapper

File: benchexec/tools/cpa-witness-exec.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    def executable(self):
+        executable = util.find_executable('cpa_witness_exec.py', 'scripts/cpa_witness_exec.py')
+        executableDir = os.path.join(os.path.dirname(executable), os.path.pardir)
+        if os.path.isdir(os.path.join(executableDir, 'src')):
+            self._buildCPAchecker(executableDir)
+        if not os.path.isfile(os.path.join(executableDir, "cpachecker.jar")):
+            logging.warning("Required JAR file for CPAchecker not found in {0}.".format(executableDir))
+        return executable
+
+    def program_files(self, executable):
```
Comment (kfriedberger) [inline benchexec/tools/cpa-witness-exec.py]: same as CPAchecker-tool-wrapper

File: benchexec/tools/cpa-witness-exec.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        installDir = os.path.join(os.path.dirname(executable), os.path.pardir)
+        return util.flatten(util.expand_filename_pattern(path, installDir) for path in self.REQUIRED_PATHS)
+
+    def _buildCPAchecker(self, executableDir):
```
Comment (kfriedberger) [inline benchexec/tools/cpa-witness-exec.py]: same as CPAchecker-tool-wrapper

File: benchexec/tools/cpa-witness-exec.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+                if not status:
+                    status = newStatus
+                elif newStatus != result.RESULT_UNKNOWN:
+                    status = "{0} ({1})".format(status, newStatus)
+
+        if not status:
+            status = result.RESULT_ERROR
+        return status
+
+    def get_value_from_output(self, lines, identifier):
```
Comment (kfriedberger) [inline benchexec/tools/cpa-witness-exec.py]: same as CPAchecker-tool-wrapper

**Comment by PhilippWendler** [comment]
@kfriedberger The name "tool wrapper" has been replaced more than a year ago by "tool-info module".

**Comment by kfriedberger** [comment]
@PhilippWendler 
I knew there was another name, but did not remember it. I have updated my comments :-)

**Comment by leostrakosch** [comment]
Thank you @kfriedberger for the detailed review.
I've tried to remove as much redundancy as possible by inheriting from the cpachecker module.
In addition, I've removed all functionality that is not needed at the moment - let me hear what you think.

File: benchexec/tools/cpa-witness-exec.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                logging.warning('Time limit already specified in command-line options,' +
+                                ' not adding time limit from benchmark definition to the command line.')
+            else:
+                added_options += ["-timelimit", str(rlimits[SOFTTIMELIMIT]) + 's']  # benchmark-xml uses seconds as unit
+
+        if "-stats" not in options:
+            added_options.append("-stats")
+
+        spec = ["-spec", propertyfile] if propertyfile is not None else []
+        added_options += spec
```
Comment (PhilippWendler) [inline benchexec/tools/cpa-witness-exec.py]: This handling of options could be refactored into a separate method in `cpachecker.py` and refactored here.

File: benchexec/tools/cpa-witness-exec.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            if line.startswith('Verification result: '):
+                line = line[21:].strip()
+                if line.startswith('FALSE'):
+                    status = result.RESULT_FALSE_REACH
+                else:
+                    status = result.RESULT_UNKNOWN
+                break
+
+        if not status:
+            status = result.RESULT_ERROR
```
Comment (PhilippWendler) [inline benchexec/tools/cpa-witness-exec.py]: Can this module give more descriptive errors like CPAchecker's module does?

For example, IIRC in SV-Comp it was a requirement to detect and warn about invalid witnesses (cf. https://github.com/sosy-lab/benchexec/commit/363eed068ad0380c9744299525f70253415861d2). Cases like `exception` are useful for debugging.

File: benchexec/tools/cpa-witness-exec.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+# prepare for Python 3
+from __future__ import absolute_import, division, print_function, unicode_literals
```
Comment (PhilippWendler) [inline benchexec/tools/cpa-witness-exec.py]: This is actually unnecessary. BenchExec only supports Python 3 except in a few specific parts.

File: benchexec/tools/cpa-witness2test.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import logging
+
+import benchexec.result as result
+import benchexec.util as util
+import benchexec.tools.cpachecker as cpachecker
+from benchexec.model import SOFTTIMELIMIT
```
Comment (PhilippWendler) [inline benchexec/tools/cpa-witness2test.py]: Now most imports are unnecessary.

File: benchexec/tools/cpa-witness2test.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+    def version(self, executable):
+        stdout = self._version_from_tool(executable, '-version')
+        return stdout
+
+    def name(self):
+        return 'CPA-witness2test'
+
+    def cmdline(self, executable, options, tasks, propertyfile=None, rlimits={}):
+        additional_options = super(Tool, self)._get_additional_options(options, propertyfile, rlimits)
```
Comment (PhilippWendler) [inline benchexec/tools/cpa-witness2test.py]: Why not simply use `self.`?

File: benchexec/tools/cpa-witness2test.py [hunk 12]
```diff
@@ -32,7 +28,7 @@ class Tool(cpachecker.Tool):
     """
 
     def executable(self):
-        super(Tool, self).executable()  # Makes sure that CPAchecker can be called, shows a warning otherwise
+        self.executable()  # Makes sure that CPAchecker can be called, shows a warning otherwise
```
Comment (PhilippWendler) [inline benchexec/tools/cpa-witness2test.py]: No, here we need `super`, right now this is endless recursion.

Comment (leostrakosch) [inline benchexec/tools/cpa-witness2test.py]: Yeah, that was too eager..

**Pull request merged**